File: src/Parser/SubjectParser.ts
import { JSDOM } from 'jsdom'

interface Topic {
    name: string,
    link: string
}

interface SubjectParserResult {
    subject: string,
    topics: Topic[]
}

interface SubjectParser {
    (subject: string, link: string): Promise<SubjectParserResult>
}

const SubjectParser: SubjectParser = async (subject, link) => {
    const $ = await JSDOM.fromURL(link)
    const document = $.window.document

    let topicsObject: SubjectParserResult = { subject, topics: []}

    document.querySelectorAll('div.sf-section > table > tbody > tr > td > li > a')
        .forEach(element => {
            topicsObject.topics.push({
                name: element.textContent,
                //@ts-ignore
                link: element.href
            })
        })

    return topicsObject
}

export default SubjectParser

File: parser/subject_parser.go
package parser

import (
	"fmt"
	"github.com/PuerkitoBio/goquery"
	"github.com/gocolly/colly"
)

type Topic struct {
	Name string
	Link string
}

type SubjectParserResult struct {
	Degree string
	Topics []Topic
}

func SubjectParser(degree string, link string) (topics SubjectParserResult, err error) {
	c := colly.NewCollector()
	topics.Degree = degree

	c.OnHTML("li", func(e *colly.HTMLElement) {
		e.DOM.Find("div.sf-section table tbody tr td li a").Each(func(_ int, selection *goquery.Selection) {
			if href, exists := selection.Attr("href"); exists {
				topics.Topics = append(topics.Topics, Topic{
					Name: selection.Text(),
					Link: href,
				})
			}
		})
	})

	c.OnScraped(func(r *colly.Response) {
		fmt.Println("Subject Scrapped", r.Request.URL)
	})

	err = c.Visit(link)

	return
}
